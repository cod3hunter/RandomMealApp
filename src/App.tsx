import React, {useState, useCallback} from 'react';
import styled from 'styled-components/native';
import {Button, MealDetails} from './components';
import {getRandomMeal} from './services/RequestService';

const Container = styled.View`
  flex: 1;
  align-items: center;
  padding: 16px;
`;

const App: React.FC = () => {
  const [meal, setMeal] = useState<Meal | Partial<Meal>>({});
  const [isLoading, setIsLoading] = useState(false);
  const generateMeal = useCallback(() => {
    (async () => {
      setIsLoading(true);
      const response = await getRandomMeal();
      setMeal(response);
      setIsLoading(false);
    })();
  }, []);
  return (
    <Container>
      <Button text="Generate random meal" onPress={generateMeal} />
      <MealDetails data={meal} isLoading={isLoading} />
    </Container>
  );
};

export default App;
