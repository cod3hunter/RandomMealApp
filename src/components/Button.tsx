import React from 'react';
import styled from 'styled-components/native';

const Container = styled.TouchableOpacity`
  width: 100%;
  elevation: 2;
  background-color: blueviolet;
  border-radius: 4px;
  padding: 8px;
  align-items: center;
`;

const Text = styled.Text`
  font-size: 20px;
  font-weight: bold;
  color: white;
`;

type Props = {
  text: string;
  onPress: () => void;
};

const Button: React.FC<Props> = ({text, onPress}) => {
  return (
    <Container {...{onPress}}>
      <Text>{text}</Text>
    </Container>
  );
};

export default Button;
