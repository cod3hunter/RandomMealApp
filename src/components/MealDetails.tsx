import React, {useState, useEffect} from 'react';
import {Dimensions, ActivityIndicator} from 'react-native';
import styled from 'styled-components/native';
import {normalizeIngredients} from '../services/utils';

const {width} = Dimensions.get('window');

const LoadingContainer = styled.View`
  width: 100%;
  height: 100%;
  align-items: center;
  justify-content: center;
`;

const Container = styled.ScrollView`
  width: 100%;
  margin-top: 32px;
`;

const Text = styled.Text`
  font-size: 16px;
  margin-top: 8px;
`;

const Image = styled.Image.attrs({height: width * 0.5, width: width * 0.8})`
  width: ${width * 0.8}px;
  height: ${width * 0.5}px;
  margin-top: 16px;
  border-radius: 4px;
`;

const Title = styled.Text`
  font-size: 24px;
  font-weight: bold;
`;

const Subtitle = styled(Title)`
  margin-top: 24px;
`;

const InstructionsText = styled(Text)`
  text-align: center;
`;

type Props = {
  data: Meal;
  isLoading?: boolean;
};

const MealDetails: React.FC<Props> = ({data}) => {
  const [ingredients, setIngredients] = useState<string[]>([]);

  useEffect(() => {
    if (data.strIngredient1) {
      setIngredients(normalizeIngredients(data));
    }
  }, [data]);

  if (ingredients.length < 1) {
    return (
      <LoadingContainer>
        <ActivityIndicator size={30} color="blueviolet" />
      </LoadingContainer>
    );
  }

  return (
    <Container contentContainerStyle={{alignItems: 'center'}}>
      <Title>{data.strMeal}</Title>
      <Image source={{uri: data.strMealThumb}} />
      {ingredients.map((ingredient, index) => {
        return <Text key={index}>{ingredient}</Text>;
      })}
      <Subtitle>Instructions</Subtitle>
      <InstructionsText>{data.strInstructions}</InstructionsText>
    </Container>
  );
};

MealDetails.defaultProps = {
  isLoading: false,
};

export default MealDetails;
