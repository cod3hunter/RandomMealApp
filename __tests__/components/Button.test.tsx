import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {Button} from '../../src/components';

it('renders correctly', () => {
  const tree = renderer
    .create(<Button text="teste" onPress={() => console.log('teste')} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
