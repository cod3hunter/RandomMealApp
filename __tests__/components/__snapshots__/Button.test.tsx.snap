// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "blueviolet",
      "borderRadius": 4,
      "elevation": 2,
      "opacity": 1,
      "paddingBottom": 8,
      "paddingLeft": 8,
      "paddingRight": 8,
      "paddingTop": 8,
      "width": "100%",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "white",
          "fontSize": 20,
          "fontWeight": "bold",
        },
      ]
    }
  >
    teste
  </Text>
</View>
`;
